packages/api/lambdas/kinesis-consumer.js

```diff
@@ -34,63 +36,73 @@ async function getKinesisRules(event) {
+  const item = {
+    workflow: kinesisRule.workflow,
+    provider: kinesisRule.provider,
+    collection: kinesisRule.collection,
+    payload: eventObject
+  };
 
-  return await Promise.all(oneTimeRulePromises);
+  return Rule.buildPayload(item)
+    .then((payload) => queueWorkflowMessage(payload));
 }
 
 /**
- * `validateMessage` validates an event as being valid for creating a workflow. See the messageSchema defined at
- * the top of this file.
+ * `validateMessage` validates an event as being valid for creating a workflow.
+ * See the messageSchema defined at the top of this file.
  *
- * @param {object} event lambda event
- * @returns {(error|object)} Throws an Ajv.ValidationError if event object is invalid. Returns the event object if event is valid.
+ * @param {Object} event - lambda event
+ * @returns {(error|Object)} Throws an Ajv.ValidationError if event object is invalid.
+ * Returns the event object if event is valid.
  */
 async function validateMessage(event) {
-  const ajv = new Ajv({allErrors: true});
+  const ajv = new Ajv({ allErrors: true });
   const validate = ajv.compile(messageSchema);
   return await validate(event);
 }
 
+/**
+ * Process data sent to a kinesis stream. Validate the data and
+ * create rules.
+ *
+ * @param {*} record - input to the kinesis stream
+ * @returns {Promise} promise
+ */
 async function processRecord(record) {
   const dataBlob = record.kinesis.data;
   const dataString = Buffer.from(dataBlob, 'base64').toString();
   const eventObject = JSON.parse(dataString);
 
-  await validateMessage(eventObject)
+  return validateMessage(eventObject)
     .then(getKinesisRules)
-    .then((kinesisRules) => {
-      return createOneTimeRules(kinesisRules);
-    });
+    .then((kinesisRules) => (
+      Promise.all(kinesisRules.map((kinesisRule) => queueMessageForRule(kinesisRule, eventObject)))
+    ));
 }
 
 /**
- * `handler` Looks up enabled 'kinesis'-type rules associated with the collection in the event argument. It
- * creates new onetime rules for each rule found to trigger the workflow defined in the 'kinesis'-type rule.
+ * `handler` Looks up enabled 'kinesis'-type rules associated with the collection
+ * in the event argument. It creates new onetime rules for each rule found to trigger
+ * the workflow defined in the 'kinesis'-type rule.
  *
- * @param {*} event lambda event
- * @param {*} context lambda context
- * @param {*} cb callback function to explicitly return information back to the caller.
+ * @param {*} event - lambda event
+ * @param {*} context - lambda context
+ * @param {*} cb - callback function to explicitly return information back to the caller.
  * @returns {(error|string)} Success message or error
  */
 function handler(event, context, cb) {
   const records = event.Records;
 
-  return Promise.all(records.map(r => processRecord(r)))
-    .then((results) => cb(null, results.filter(r => r !== undefined)))
+  return Promise.all(records.map((r) => processRecord(r)))
```

Comment:
I think we can make this even terser - but it's a style preference:
```javascript
return Promise.all(records.map(processRecord))
```

Reply:
Fixed